library/nist_kw.c

```diff
@@ -0,0 +1,689 @@
+    mbedtls_platform_zeroize( outbuff, KW_SEMIBLOCK_LENGTH * 2 );
+    mbedtls_cipher_finish( &ctx->cipher_ctx, NULL, &olen );
+    return( ret );
+}
+
+/*
+ * W-1 function as defined in RFC 3394 section 2.2.2:
+ */
+static int unwrap( mbedtls_nist_kw_context *ctx,
+                   const unsigned char *input, size_t semiblocks,
+                   unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    const size_t s = 6 * ( semiblocks -1 );
+    size_t olen, i;
+    uint64_t t = 0;
+    unsigned char outbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char inbuff[KW_SEMIBLOCK_LENGTH * 2];
+    unsigned char *R = output + ( semiblocks -1 ) * KW_SEMIBLOCK_LENGTH ;
+    unsigned char *A = output ;
+
+    if( semiblocks < MIN_SEMIBLOCKS_COUNT )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
+    memcpy( output, input, semiblocks * KW_SEMIBLOCK_LENGTH );
+
+    /* Calculate intermediate values */
+    for( t = s; t >= 1; t-- )
+    {
+        for( i = 0; i < sizeof( t ); i++ )
+        {
+            A[i] ^= ( unsigned char )( ( t >> ( ( sizeof( t ) - 1 - i ) * 8 ) )
+                    & 0xff );
+        }
+        memcpy( inbuff, A, KW_SEMIBLOCK_LENGTH );
+        memcpy( inbuff + KW_SEMIBLOCK_LENGTH, R, KW_SEMIBLOCK_LENGTH );
+
+        ret = mbedtls_cipher_update( &ctx->cipher_ctx,
+                                     inbuff, 16, outbuff, &olen);
+        if( ret != 0 )
+            goto cleanup;
+
+        memcpy( A, outbuff, KW_SEMIBLOCK_LENGTH );
+
+        /* Set R as LSB64 of outbuff */
+        memcpy( R, outbuff + KW_SEMIBLOCK_LENGTH, KW_SEMIBLOCK_LENGTH );
+        R -= KW_SEMIBLOCK_LENGTH;
+        if( R < output + KW_SEMIBLOCK_LENGTH )
+            R = output + ( semiblocks -1 ) * KW_SEMIBLOCK_LENGTH;
+
+    }
+
+    *out_len = semiblocks * KW_SEMIBLOCK_LENGTH;
+
+cleanup:
+    if( ret != 0)
+        mbedtls_platform_zeroize( output, semiblocks * KW_SEMIBLOCK_LENGTH );
+    mbedtls_platform_zeroize( inbuff, KW_SEMIBLOCK_LENGTH * 2 );
+    mbedtls_platform_zeroize( outbuff, KW_SEMIBLOCK_LENGTH * 2 );
+
+    return( ret );
+}
+
+/*
+ * KW-AD as defined in SP 800-38F section 6.2
+ * KWP-AD as defined in SP 800-38F section 6.3
+ */
+int mbedtls_nist_kw_unwrap( mbedtls_nist_kw_context *ctx,
+                            mbedtls_nist_kw_mode_t mode,
+                            const unsigned char *input, size_t in_len,
+                            unsigned char *output, size_t* out_len )
+{
+    int ret = 0;
+    size_t i;
+    const uint32_t ICV1[] = NIST_KW_ICV1;
+    const uint32_t ICV2[] = NIST_KW_ICV2;
+
+    if( mode == MBEDTLS_KW_MODE_KW )
+    {
+        /*
+         * According to SP800-38F Table 1, the ciphertext length for KW
+         * must be between 3 to 2^54 semiblocks inclusive.
+         */
+        if( in_len < 24
+#if SIZE_MAX > 0x200000000000000
+            || in_len > 0x200000000000000
+#endif
+            )
+        {
+            *out_len = 0;
```

Comment:
Rather than set `*out_len` before returning, put `*out_len = 0` at the beginning of the function, and assign again on success. This way you'll need fewer places with `*out_len = 0` and you won't forget some places (e.g. if `unwrap` fails just below).

Reply:
But then, in case of failure, when we `goto cleanup` we will crase, because of:
` mbedtls_platform_zeroize( output, *out_len - 1);` and `*out_len` is now set to zero

Reply:
Er, I'd missed that, but how could `*out_len - 1` be right? Why would you not zeroize the last byte?

I'd expect the logic to be either:

* `*out_len` remains set to the number of bytes written throughout the function. On error, you end up wiping what has been written so far.
* `*out_len` is set to 0 initially, and to the number of bytes written on success. On error, you wipe the whole output buffer.

Reply:
I tried the second approach, and since the output buffer should be at least semiblocks * 8 ( the assignment for `out_len`) , then this is what I tried. Can't recall why the `-1` part though

Reply:
how do you expect to calculate the bytes written on the fly? It can be done, but it adds complexity, as we write on the same position several times